src/triage/component/catwalk/evaluation.py

```diff
@@ -438,6 +456,92 @@ def needs_evaluations(self, matrix_store, model_id, subset_hash=""):
+            labels:
+            tie_breaker:
+            subset_hash:
+            bias_config:
+            matrix_type:
+            evaluation_start_time:
+            evaluation_end_time:
+            matrix_uuid:
+
+        Returns:
+
+        """
+        bias_audits = []
+        protected_df['model_id'] = model_id
+        protected_df['score'] = predictions_proba
+        protected_df['label_value'] = labels
+        df, attr_cols_input = preprocess_input_df(protected_df)
+        g = Group()
+        score_thresholds = {}
+        score_thresholds['rank_abs'] = bias_config['thresholds'].get('top_n', [])
+        score_thresholds['rank_pct'] = bias_config['thresholds'].get('percentiles', [])
+        groups_model, attr_cols = g.get_crosstabs(df,
+                                                  score_thresholds=score_thresholds,
+                                                  model_id=model_id,
+                                                  attr_cols=attr_cols_input)
+        b = Bias()
+        ref_groups_method = bias_config.ref_groups_method
+        if ref_groups_method == 'predefined' and bias_config.ref_groups:
+            bias_df = b.get_disparity_predefined_groups(groups_model, df, bias_config.ref_groups)
+        elif ref_groups_method == 'majority':
+            bias_df = b.get_disparity_major_group(groups_model, df)
+        else:
+            bias_df = b.get_disparity_min_metric(groups_model, df)
+        f = Fairness(tau=bias_config.fairness_threshold)
+        group_value_df = f.get_group_value_fairness(bias_df)
+        group_value_df['subset_hash'] = subset_hash
+        group_value_df['tie_breaker'] = tie_breaker
+        group_value_df['evaluation_start_time'] = evaluation_start_time
+        group_value_df['evaluation_end_time'] = evaluation_end_time
+        group_value_df['matrix_uuid'] = matrix_uuid
+        group_value_df['parameter'] =
+        #delete score_thresholds
+        if group_value_df.empty:
+            raise ValueError("""
+            Bias audit: aequitas_audit() failed. Returned empty dataframe for model_id = {model_id}, and subset_hash = {subset_hash}
+            and predictions_schema = {schema}""".format())
+        with scoped_session(self.db_engine) as session:
+            for index, row in group_value_df.iterrows():
+                session.query(matrix_type.aequitas_obj).filter_by(
+                    model_id=model_id,
+                    evaluation_start_time=evaluation_start_time,
+                    evaluation_end_time=evaluation_end_time,
+                    subset_hash=subset_hash,
+                    parameter=parameter,
+                    tie_breaker=tie_breaker,
+                    matrix_uuid=matrix_uuid,
+                    attribute_name=row['attribute_name'],
+                    attribute_value=row['attribute_value']
+                ).delete()
+            session.bulk_insert_mappings(matrix_type.aequitas_obj, group_value_df.to_dict(orient="records"))
```

Comment:
(I know we talked about this so you probably did try, but), did you try simply `bulk_update_mappings`? _E.g._:

```python
with scoped_session(self.db_engine) as session:
    session.bulk_update_mappings(matrix_type.aequitas_obj, group_value_df.to_dict(orient="records"))
```

It looks like your database index isn't set on the dataframe; but, if it were:

```python
with scoped_session(self.db_engine) as session:
    session.bulk_update_mappings(matrix_type.aequitas_obj, group_value_df.reset_index().to_dict(orient="records"))
```

Of course, I presume this requires that an index is set on the database table, (and that SQLAlchemy is aware of it).

Reply:
I would _guess_ that `bulk_update_mappings` is the most efficient for the situation, or at least more so than non-bulk SQLAlchemy methods. But, if it's really a lot of data, and otherwise I suppose for the record, you might do something like:

```python
with db_engine.begin() as conn:
    # create temp table to which to COPY
    # using schema (not rows) of true target table
    conn.execute(f'''\
        create temp table group_values
        as select * from {matrix_type.aequitas_obj.__table__} limit 0
    ''')

    group_value_df.pg_copy_to('group_values', conn)

    conn.execute(f"""\
        insert into {matrix_type.aequitas_obj.__table__}
        on conflict do update
        select * from group_values
    """)
```